EpiserverRedirects.EntityFramework/Repository/RedirectRulesRepository.cs

```diff
@@ -0,0 +1,138 @@
+﻿using Forte.EpiserverRedirects.Model.RedirectRule;
+using Forte.EpiserverRedirects.Repository;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+
+namespace Forte.EpiserverRedirects.EntityFramework.Repository
+{
+    public class RedirectRulesRepository<TDbContext> : IRedirectRuleRepository
+        where TDbContext : IRedirectRulesDbContext
+    {
+        private readonly IRedirectRulesDbContext _dbContext;
+        private readonly IRedirectRuleMapper _mapper;
+
+        public RedirectRulesRepository(
+            IRedirectRulesDbContext dbContext,
+            IRedirectRuleMapper mapper)
+        {
+            _dbContext = dbContext;
+            _mapper = mapper;
+        }
+
+        public RedirectRuleModel GetById(Guid id)
+        {
+            var entity = _dbContext.RedirectRules.FirstOrDefault(rule => rule.Id == id);
+            return _mapper.MapToModel(entity);
+        }
+
+        public IList<RedirectRuleModel> GetAll()
+        {
+            return _dbContext.RedirectRules
+                .Select(entity => _mapper.MapToModel(entity))
+                .ToList();
+        }
+
+        public IList<RedirectRuleModel> Query(out int allRedirectsCount, RedirectRuleQuery query)
+        {
+            return _dbContext.RedirectRules
+                .ApplyQuery(out allRedirectsCount, query)
+                .Select(entity => _mapper.MapToModel(entity))
+                .ToList();
+        }
+
+        public IList<RedirectRuleModel> GetByContent(IList<int> contentIds)
+        {
+            return _dbContext.RedirectRules
+                .Where(x => x.ContentId.HasValue && contentIds.Contains(x.ContentId.Value))
+                .Select(entity => _mapper.MapToModel(entity))
+                .ToList();
+        }
+
+        public RedirectRuleModel FindRegexMatch(string patern)
```

Comment:
I think that we should not mix finding a match and querying, as they are serving different purposes.
Find*Match() returns only one matched item based on requested URL (no query). Matched item is supposed to be cached in a url-Item dictionary.

The Query() method used in UI or some third party code which may use it. We can extend it to eliminate the repo.GetByContent method.

Reply:
`Query` is generic enough that does not necessarily have to be used ONLY when we want to retrieve multiple items.
I don't see any issues in using `Query` in the following scenario as well: 'I want to find all rules that match my URL and I want them to be ordered by priority'. And after you get them you simply call `FirstOrDefault` e.g. in the `ExactMatchResolver` and thanks to that we have data access separated from our business logic.

And taking into consideration that this interface will have to be reimplemented by the custom repository's authors I wanted to reduce the number of methods that are present here.

Reply:
No problem with doing all Reads through Query. The problem is how we save it in cache. For finding a Match - we use just url Pattern as a Key, and store a Match in cache by that Key (patern)..
If we extract a match using Query - we will have to somehow stringify the Query object into some string key. It Will be a long string with all Query fields and sortings. Is it ok ?

Reply:
I thought that we agreed on removing this layer of cache as it is hardly useful right now (if not then sorry, it might have been my left and right parts of my brain agreeing on that :D).
So anyway, I would remove this cache and just leave one which caches the resolvers' result.